Incident Management Doc

jjasghar opened the pull request:
This is the first pass at created an incident management process for InstructLab. We will have problems and this is the beginning of the way to communicate them out.

Thanks to @coderanger for the inspiration for this.

leseb commented on docs/incident-management.md:
```diff
@@ -0,0 +1,123 @@
+
+## Attribution
+Before we start, I would like to mention that this document and process is inspired
+by and influenced heavily from [here][gist] originally authored by [coderanger][ranger].
```
Can't we just add this person as a Co-author to this contribution instead of mentioning this?

jjasghar replied:
No idea, I figured that there would be _major_ changes to this document as we go forward. This was just a start and figured it would be worth just "tipping the hat to them."

lhawthorn replied:
I am personally a huge fan of deeply explicit acknowledgements for many reasons, not the least of which is showing people we actually pay attention to the state of the art and best practices other devs/projects are using. JMHO.

leseb commented on docs/incident-management.md:
```diff
@@ -0,0 +1,123 @@
+
+With the multiple systems that come together for the open source project InstructLab
+to work, we need a catalogue of major issues and problems so we can track and find
+critical bugs in the open. Bulding a formalized incident management process is
```
s/Bulding/Building/

leseb commented on docs/incident-management.md:
```diff
@@ -0,0 +1,123 @@
+
+### What counts as an incident?
+
+Anything with customer or community visible negative consequences. In most cases this
```
Customer? This sounds a bit strange given this is an open source project. "Community" seems sufficient.

mairin replied:
agreed

lhawthorn replied:
+1 on community is sufficient. Customers are community members, too.

leseb reviewed (commented):
This is the first time I've seen something so formalized for an open source project. It feels very company-oriented, but I might be missing something. Another point is the operational nature of the Incident Commander (IC) role as described. Currently, InstructLab, at least in terms of the CLI, doesn't have any long-running services, so I'm not sure if we would experience outages in the same way. However, I do agree that having a process for post-mortem analysis is valuable for addressing issues. I have mixed feelings about this but want to hear from others as well :). Thanks for the write-up!

jjasghar commented:
> This is the first time I've seen something so formalized for an open source project. It feels very company-oriented, but I might be missing something. Another point is the operational nature of the Incident Commander (IC) role as described. Currently, InstructLab, at least in terms of the CLI, doesn't have any long-running services, so I'm not sure if we would experience outages in the same way. However, I do agree that having a process for post-mortem analysis is valuable for addressing issues. I have mixed feelings about this but want to hear from others as well :). Thanks for the write-up!

Ah, see, it turns out we do have a _bunch_ of back-end services required for us to run this project. We have a team of backend engineers ensuring the model training and tuning is done. This requires communication between two different groups of people, and we have no formal way to communicate. (hence this process) 
We had a "major" incident, just this last week and when I asked for a some information about it, we noticed we didn't have a plan for these types of situations.

russellb commented on docs/incident-management.md:
```diff
@@ -0,0 +1,123 @@
+
+Anything with customer or community visible negative consequences. In most cases this
+will be an outage or downtime event but some non-outage incidents include severe
+performance degradations and security events.
```
I'm having a hard time understanding what would qualify as an incident under this proposal.  Could you come up with some concrete examples?

It seems to be oriented toward running a service where some level of availability is to be expected. That's not really true here. There's stuff running, but not in service to a public group of some kind.

Before commenting further, it would help if we had a sample set of scenarios to use as context for discussion.

russellb replied:
I just nonticed I'm duplicating some discussion between you and @leseb - sorry about that.

> Ah, see, it turns out we do have a bunch of back-end services required for us to run this project. We have a team > of backend engineers ensuring the model training and tuning is done. This requires communication between two > different groups of people, and we have no formal way to communicate. (hence this process)
> We had a "major" incident, just this last week and when I asked for a some information about it, we noticed we didn't have a plan for these types of situations.

but none of this is visible to the public right now, so I'm not sure a public process like this makes sense.

lhawthorn replied:
@russellb I take your comment to mean "once our backend services like model training are publicly visible, we should have a process for dealing with incidents"

I agree with you!

However, as we did have a hiccup that caused @jjasghar to open this PR, should we at least have some documented process for dealing with said hiccups amongst the maintainer team? I am sure we have one that is informal and ad hoc, but I'd like to see it written down (for my ignorant self).

I think @jjasghar can maybe just give me a summary of the what happened and how to remediate/deal in this issue. The deep specifics are unlikely to be interesting or useful for GitHub history purposes. :D

russellb replied:
I don't know anything about the example in question. It would probably help to speak about it more concretely instead of in the hypothetical.

hickeyma reviewed (commented):
I am in agreement with @leseb and @russellb. Some of my thoughts:

- This seems like a process when you have service(s) that you support. Will the open source community be providing such support?
- The backend services I would think will remain for the most part to run the workflow to perform model alignment and tuning of a new model to be published. Do we envision exposing this externally?
- Processing of taxononmy PRs should be treated like any open source contribution: push, CI/CD, review/re-commit cycle and then merge or reject. Any capabilities run (e.g. SDG, training) for the PR should be treated like CI/CD in all projects - raise a bug if the gate is broken.
- What was the issue(s) that occurred that instigated this proposal? This is important in the context if the proposal is required or not.

nathan-weinberg commented:
@jjasghar what's the status of this doc?

jjasghar commented:
Haven't touched this since we created it. When we start hosting some level of service for our community we will need some standard procedure, but I think we can have this sit around for the time being.
